### DISCUSSION TITLE
How to filter on type of object?

### DISCUSSION THREAD

**Discussion by gmermoud** [geldata/gel#5333] [Q&A] [answered]
Hello,

I have a schema where four distinct types A, B, C, D inherit from the same abstract type X.

Now, I have a query of the form:

```
select X {
   field1,
   field2
} 
```

But I would like to filter out objects of class D, such that:

```
select X {
   field1,
   field2
} filter . is (A|B|C)
```

What is the correct way to achieve this?

**Comment by gmermoud** [comment]
OK, that's rubber duck debugging at its best. Upon hitting "send", I immediately thought of:

```
select X {
   field1,
   field2
}[is A]
```

which works. 

However, the following seems to be unsupported (I get the error `QueryError: complex type expressions are not supported here`):

```
select X {
   field1,
   field2
}[is (A|B|C)]
```

What is the correct way to achieve what I want then?

**Comment by raddevon** [comment]
Update: This method from @elprans is better than the one I suggested:
> `filter X is (A | B | C)` should work and is the idiomatic way of filtering by type. Filtering by `__type__.name` works, but is somewhat less efficient.

Try this:
```
select X {
  field1,
  field2
} filter .__type__.name = 'default::A'
or .__type__.name = 'default::B'
or .__type__.name = 'default::C';
```
assuming the types `A`, `B`, and `C` are all in the `default` module.

**Comment by gmermoud** [comment]
This works, thanks a lot!

I suppose the following syntax would be nice:

```
select X {
  field1,
  field2
} filter . is (A|B|C)
```
Assuming that . is indeed a shorthand for the object itself.

Thanks a lot for your help.

Greg

**Reply by elprans** [reply]
`filter X is (A | B | C)` should work and is the idiomatic way of filtering by type. Filtering by `__type__.name` works, but is somewhat less efficient.

**Reply by gmermoud** [reply]
@elprans thanks a lot! Indeed, I had missed this construct.

**Comment by gmermoud** [comment, marked as answer]
Actually, based on @elprans's reply above, the correct way to achieve this is:

```
select X {
  field1,
  field2
} filter X is (A | B | C)
```

The following also works:

```
select X {
  field1,
  field2
} filter X is not D
```
Thanks all!
